File: pay_calculator.py
from flask import Blueprint, render_template, jsonify
import json
import os
from datetime import datetime, time

pay_calc_bp = Blueprint('pay_calc', __name__)

def load_config():
    """Load MA00020 configuration file for the pay calculator"""
    try:
        config_path = os.path.join('data', 'ma00020_config.json')
        with open(config_path, 'r', encoding='utf-8') as f:
            award_config = json.load(f)
        return award_config
    except Exception as e:
        print(f"Error loading MA00020 configuration: {str(e)}")
        return None

def get_classification_rate(classification, level=None):
    """Get the hourly rate for a classification and level"""
    config = load_config()
    if not config:
        return None
    
    if classification == "CW/ECW 1" and level:
        return config['classifications']['CW/ECW 1']['levels'][level]['hourly_rate']
    else:
        return config['classifications'][classification]['hourly_rate']

def calculate_overtime_pay(base_rate, hours, overtime_type):
    """Calculate overtime pay based on type and hours"""
    config = load_config()
    if not config:
        return 0
    
    if overtime_type == 'weekday':
        first_two = min(hours, 2)
        remaining = max(0, hours - 2)
        rates = config['overtime']['weekday']
        return (first_two * base_rate * rates['first_two_hours']['rate_multiplier'] +
                remaining * base_rate * rates['after_two_hours']['rate_multiplier'])
    else:
        multiplier = config['overtime'][overtime_type]['rate_multiplier']
        return hours * base_rate * multiplier

def calculate_allowances(allowance_type, conditions=None):
    """Calculate applicable allowances"""
    config = load_config()
    if not config:
        return 0
    
    allowances = config['allowances']
    if allowance_type == 'industry':
        return allowances['industry']['general_building']['amount']
    elif allowance_type == 'tools':
        # Default to carpenter rate if not specified
        return allowances['tools']['carpenter_joiner_stonemason_tilelayer']['amount']
    elif allowance_type == 'meal':
        if conditions and 'overtime_required' in conditions:
            return allowances['meal']['overtime']['amount']
    elif allowance_type == 'travel':
        if conditions and 'distance' in conditions:
            distance = conditions['distance']
            if distance >= allowances['travel']['min_distance']:
                return distance * allowances['travel']['rate_per_km']
    return 0

@pay_calc_bp.route('/pay_calculator')
def index():
    """Render the pay calculator page"""
    try:
        # Load MA00020 configuration
        award_config = load_config()
        
        if not award_config:
            return render_template('error.html', 
                                 error="Failed to load MA00020 configuration",
                                 details="Please check if the configuration file exists and is valid.")
        
        return render_template('pay_calculator.html',
                             award_config=award_config)
                             
    except Exception as e:
        print(f"Error in pay calculator route: {str(e)}")
        return render_template('error.html',
                             error="An error occurred while loading the pay calculator",
                             details=str(e))

@pay_calc_bp.route('/calculate', methods=['POST'])
def calculate():
    """Calculate pay based on form submission"""
    try:
        data = request.get_json()
        
        # Get base rate
        classification = data.get('classification')
        level = data.get('level') if classification == "CW/ECW 1" else None
        base_rate = get_classification_rate(classification, level)
        
        if not base_rate:
            return jsonify({"error": "Invalid classification"}), 400
        
        # Get employment type and apply loading
        employment_type = data.get('employmentType')
        config = load_config()
        loading = config['employment_types'][employment_type]['loading']
        base_rate *= (1 + loading)
        
        # Calculate base pay
        ordinary_hours = float(data.get('ordinaryHours', 0))
        base_pay = base_rate * ordinary_hours
        
        # Calculate overtime
        overtime_hours = float(data.get('overtimeHours', 0))
        weekend_hours = float(data.get('weekendHours', 0))
        public_holiday_hours = float(data.get('publicHolidayHours', 0))
        
        overtime_pay = calculate_overtime_pay(base_rate, overtime_hours, 'weekday')
        weekend_pay = calculate_overtime_pay(base_rate, weekend_hours, 'weekend')
        public_holiday_pay = calculate_overtime_pay(base_rate, public_holiday_hours, 'public_holiday')
        
        # Calculate allowances
        tool_allowance = calculate_allowances('tools')
        industry_allowance = calculate_allowances('industry')
        meal_allowance = calculate_allowances('meal', {'overtime_required': overtime_hours > 0})
        travel_distance = float(data.get('travelDistance', 0))
        travel_allowance = calculate_allowances('travel', {'distance': travel_distance})
        
        # Calculate total
        total_pay = (base_pay + overtime_pay + weekend_pay + public_holiday_pay +
                    tool_allowance + industry_allowance + meal_allowance + travel_allowance)
        
        return jsonify({
            "success": True,
            "data": {
                "base_pay": round(base_pay, 2),
                "overtime_pay": round(overtime_pay, 2),
                "weekend_pay": round(weekend_pay, 2),
                "public_holiday_pay": round(public_holiday_pay, 2),
                "tool_allowance": round(tool_allowance, 2),
                "industry_allowance": round(industry_allowance, 2),
                "meal_allowance": round(meal_allowance, 2),
                "travel_allowance": round(travel_allowance, 2),
                "total_pay": round(total_pay, 2)
            }
        })
        
    except Exception as e:
        return jsonify({
            "success": False,
            "error": str(e)
        }), 500 

File: static/js/payCalculator.js
// Base rates for each classification level
const CLASSIFICATION_RATES = {
    'CW1': 25.41,
    'CW2': 26.64,
    'CW3': 27.87,
    'CW4': 29.10
};

// Allowance amounts
const ALLOWANCES = {
    tool: 21.71,      // per week
    industry: 1.15,   // per hour
    meal: 17.55,      // per shift
    travel: 0.78      // per km
};

// Initialize form elements
document.addEventListener('DOMContentLoaded', function() {
    const form = document.getElementById('payCalculatorForm');
    const employeeNameInput = document.querySelector('.employee-name-input');
    const exportBtn = document.getElementById('exportBtn');
    const classificationSelect = document.getElementById('classification');
    const levelSelect = document.getElementById('level');
    const employmentTypeSelect = document.getElementById('employmentType');
    const agreedHoursGroup = document.getElementById('agreedHoursGroup');
    const shiftworkSelect = document.getElementById('shiftwork');
    const cribBreakGroup = document.getElementById('cribBreakGroup');
    const toolAllowanceSelect = document.getElementById('toolAllowance');
    const tradeGroup = document.getElementById('tradeGroup');
    const averagingArrangementSelect = document.getElementById('averagingArrangement');
    const averagingPeriodGroup = document.getElementById('averagingPeriodGroup');
    const weeklyHoursInput = document.getElementById('weeklyHours');

    // Handle classification change
    classificationSelect.addEventListener('change', function() {
        levelSelect.style.display = this.value === 'CW/ECW' ? 'block' : 'none';
    });

    // Handle employment type change
    employmentTypeSelect.addEventListener('change', function() {
        agreedHoursGroup.style.display = this.value === 'part-time' ? 'block' : 'none';
    });

    // Handle shiftwork change
    shiftworkSelect.addEventListener('change', function() {
        cribBreakGroup.style.display = this.value === 'yes' ? 'block' : 'none';
    });

    // Handle tool allowance change
    toolAllowanceSelect.addEventListener('change', function() {
        tradeGroup.style.display = this.value === 'yes' ? 'block' : 'none';
    });

    // Handle averaging arrangement change
    averagingArrangementSelect.addEventListener('change', function() {
        averagingPeriodGroup.style.display = this.value === 'yes' ? 'block' : 'none';
    });

    // Validate weekly hours
    weeklyHoursInput.addEventListener('input', function() {
        const hours = parseFloat(this.value);
        const isAveraging = averagingArrangementSelect.value === 'yes';
        const maxHours = isAveraging ? 152 : 38; // 38 hours * 4 weeks for averaging

        if (hours > maxHours) {
            this.setCustomValidity(`Maximum hours cannot exceed ${maxHours} for ${isAveraging ? 'averaging period' : 'week'}`);
        } else {
            this.setCustomValidity('');
        }
    });

    // Validate break times
    function validateBreaks() {
        const ordinaryHours = parseFloat(document.getElementById('ordinaryHours').value) || 0;
        const mealBreak = document.getElementById('mealBreak').value;
        const restBreak = document.getElementById('restBreak').value;
        const cribBreak = document.getElementById('cribBreak').value;
        const isShiftwork = shiftworkSelect.value === 'yes';

        if (ordinaryHours > 5 && mealBreak === 'no') {
            document.getElementById('mealBreak').setCustomValidity('Meal break required after 5 hours');
        } else {
            document.getElementById('mealBreak').setCustomValidity('');
        }

        if (ordinaryHours > 4 && restBreak === 'no') {
            document.getElementById('restBreak').setCustomValidity('Rest break required after 4 hours');
        } else {
            document.getElementById('restBreak').setCustomValidity('');
        }

        if (isShiftwork && cribBreak === 'no') {
            document.getElementById('cribBreak').setCustomValidity('Crib break required for shift work');
        } else {
            document.getElementById('cribBreak').setCustomValidity('');
        }
    }

    // Add break validation to form submission
    form.addEventListener('submit', function(e) {
        validateBreaks();
        if (!form.checkValidity()) {
            e.preventDefault();
        }
    });

    // Handle employee name input
    employeeNameInput.addEventListener('input', function() {
        // You can add validation or auto-save functionality here
    });

    // Handle export button
    exportBtn.addEventListener('click', function() {
        exportPaySummary();
    });
});

async function calculatePay() {
    try {
        // Get form values
        const formData = {
            classification: document.getElementById('classification').value,
            level: document.getElementById('level').value,
            employmentType: document.getElementById('employmentType').value,
            agreedHours: parseFloat(document.getElementById('agreedHours').value) || 0,
            shiftwork: document.getElementById('shiftwork').value,
            shiftType: document.getElementById('shiftType').value,
            workLocation: document.getElementById('workLocation').value,
            ordinaryHours: parseFloat(document.getElementById('ordinaryHours').value) || 0,
            overtimeHours: parseFloat(document.getElementById('overtimeHours').value) || 0,
            weekendHours: parseFloat(document.getElementById('weekendHours').value) || 0,
            publicHolidayHours: parseFloat(document.getElementById('publicHolidayHours').value) || 0,
            confinedSpace: document.getElementById('confinedSpace').value,
            workingAtHeights: document.getElementById('workingAtHeights').value,
            extremeWeather: document.getElementById('extremeWeather').value,
            toolAllowance: document.getElementById('toolAllowance').value,
            industryAllowance: document.getElementById('industryAllowance').value,
            mealAllowance: document.getElementById('mealAllowance').value,
            travelDistance: parseFloat(document.getElementById('travelDistance').value) || 0,
            overtimeAgreement: document.getElementById('overtimeAgreement').value,
            weekendAgreement: document.getElementById('weekendAgreement').value,
            publicHolidayAgreement: document.getElementById('publicHolidayAgreement').value
        };

        // Basic validation
        if (!formData.classification || !formData.employmentType) {
            alert('Please select classification and employment type');
            return;
        }

        if (formData.classification === 'CW/ECW 1' && !formData.level) {
            alert('Please select a level for CW/ECW 1 classification');
            return;
        }

        if (formData.employmentType === 'part_time' && !formData.agreedHours) {
            alert('Please enter agreed hours for part-time employment');
            return;
        }

        if (formData.shiftwork === 'yes' && !formData.shiftType) {
            alert('Please select shift type for shiftwork');
            return;
        }

        if (formData.ordinaryHours < 0 || formData.overtimeHours < 0 || 
            formData.weekendHours < 0 || formData.publicHolidayHours < 0) {
            alert('Hours cannot be negative');
            return;
        }

        // Validate agreements
        if (formData.overtimeHours > 0 && formData.overtimeAgreement === 'no') {
            alert('Written agreement required for overtime work');
            return;
        }

        if (formData.weekendHours > 0 && formData.weekendAgreement === 'no') {
            alert('Written agreement required for weekend work');
            return;
        }

        if (formData.publicHolidayHours > 0 && formData.publicHolidayAgreement === 'no') {
            alert('Written agreement required for public holiday work');
            return;
        }

        // Send data to server for calculation
        const response = await fetch('/pay_calculator/calculate', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(formData)
        });

        const result = await response.json();

        if (!result.success) {
            throw new Error(result.error);
        }

        // Update displays
        updateDisplays(result.data);

    } catch (error) {
        console.error('Error calculating pay:', error);
        alert('An error occurred while calculating pay. Please try again.');
    }
}

function updateDisplays(amounts) {
    // Update header total
    document.getElementById('headerTotalAmount').textContent = formatCurrency(amounts.total_pay);

    // Update summary panel
    document.getElementById('basePay').textContent = formatCurrency(amounts.base_pay);
    document.getElementById('overtimePay').textContent = formatCurrency(amounts.overtime_pay);
    document.getElementById('weekendPay').textContent = formatCurrency(amounts.weekend_pay);
    document.getElementById('publicHolidayPay').textContent = formatCurrency(amounts.public_holiday_pay);
    document.getElementById('shiftLoading').textContent = formatCurrency(amounts.shift_loading);
    document.getElementById('industryAllowance').textContent = formatCurrency(amounts.industry_allowance);
    document.getElementById('toolAllowance').textContent = formatCurrency(amounts.tool_allowance);
    document.getElementById('mealAllowance').textContent = formatCurrency(amounts.meal_allowance);
    document.getElementById('travelAllowance').textContent = formatCurrency(amounts.travel_allowance);
    document.getElementById('specialConditions').textContent = formatCurrency(amounts.special_conditions);
    document.getElementById('totalPay').textContent = formatCurrency(amounts.total_pay);
}

function formatCurrency(amount) {
    return '$' + amount.toFixed(2);
}

function exportPaySummary() {
    const employeeName = document.querySelector('.employee-name-input').value || 'Employee';
    const classification = document.getElementById('classification').value;
    const level = document.getElementById('level').value;
    const employmentType = document.getElementById('employmentType').value;
    const shiftwork = document.getElementById('shiftwork').value;
    const shiftType = document.getElementById('shiftType').value;
    const workLocation = document.getElementById('workLocation').value;
    
    // Create CSV content
    const csvContent = [
        ['Pay Summary for ' + employeeName],
        ['Classification', classification + (level ? ` - Level ${level}` : '')],
        ['Employment Type', employmentType],
        ['Work Pattern', shiftwork === 'yes' ? `Shiftwork - ${shiftType}` : 'Standard'],
        ['Work Location', workLocation],
        [''],
        ['Component', 'Amount'],
        ['Base Pay', document.getElementById('basePay').textContent],
        ['Overtime', document.getElementById('overtimePay').textContent],
        ['Weekend Hours', document.getElementById('weekendPay').textContent],
        ['Public Holiday', document.getElementById('publicHolidayPay').textContent],
        ['Shift Loading', document.getElementById('shiftLoading').textContent],
        ['Industry Allowance', document.getElementById('industryAllowance').textContent],
        ['Tool Allowance', document.getElementById('toolAllowance').textContent],
        ['Meal Allowance', document.getElementById('mealAllowance').textContent],
        ['Travel Allowance', document.getElementById('travelAllowance').textContent],
        ['Special Conditions', document.getElementById('specialConditions').textContent],
        [''],
        ['Total Pay', document.getElementById('totalPay').textContent]
    ].map(row => row.join(',')).join('\n');

    // Create and trigger download
    const blob = new Blob([csvContent], { type: 'text/csv;charset=utf-8;' });
    const link = document.createElement('a');
    link.href = URL.createObjectURL(blob);
    link.download = `pay_summary_${employeeName.toLowerCase().replace(/\s+/g, '_')}.csv`;
    link.click();
} 